# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/flyrank-bih/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

### I chose Lane 4: CTR/Engagement Opportunity Scoring. From the Week 1 exploration I conducted on the starter data, I found that CTR is heavily influenced by position (the average CTR on page_1 is around 0.35, while in the deep pages it’s only about 0.055)—so comparing CTRs across pages without accounting for their position can be misleading. What’s more interesting is that when I compared CTRs across content_types at the same position tier, comparison articles consistently had the lowest CTR in nearly all tiers with sufficient data (page_1, page_3_5, striking)—meaning it’s not just about position; there’s something about the format or relevance of this content type that causes it to underperform compared to other types at equivalent ranks. This finding aligns with the core question of Lane 4: which pages are under-capturing clicks relative to their position, and are worth reviewing for their title, meta tags, and structure. Compared to other lanes with broader scopes (Lane 2 covers many types of actions at once), Lane 4 feels more focused and achievable for 7 weeks because I already have a concrete starting point: the columns ctr, avg_position, position_tier, and content_type have proven to contain signals worth exploring further.

I chose Lane 4: CTR/Engagement Opportunity Scoring. From the Week 1 exploration I conducted on the starter data, I found that CTR is heavily influenced by position (the average CTR on page_1 is around 0.35, while in the deep pages it’s only about 0.055)—so comparing CTRs across pages without accounting for their position can be misleading. What’s more interesting is that when I compared CTRs across content_types at the same position tier, comparison articles consistently had the lowest CTR in nearly all tiers with sufficient data (page_1, page_3_5, striking)—meaning it’s not just about position; there’s something about the format or relevance of this content type that causes it to underperform compared to other types at equivalent ranks. This finding aligns with the core question of Lane 4: which pages are under-capturing clicks relative to their position, and are worth reviewing for their title, meta tags, and structure. Compared to other lanes with broader scopes (Lane 2 covers many 

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

### The unit of analysis in this project is a single content page (content_id), based on its 90-day performance metrics (impressions, clicks, CTR, avg_position). The decision we aim to support: the content/SEO team has limited capacity to review pages one by one, so they need a priority list—which pages are most worth checking first because their CTR is significantly below expectations for their position. Actions that can be taken based on this output: reviewers open the top pages on the list and decide whether to rewrite the title or meta description, fix the snippet structure, or adjust the content to better match search intent—these are not automated actions, but recommendations that still require manual verification before being implemented. The cost of incorrect recommendations works both ways. If the system recommends a page that actually has no issues (false positive), the cost is the reviewer’s wasted time checking a page that doesn’t need to be changed—and in the worst-case scenario, a page that’s already performing well gets rewritten without a valid reason, risking a decline in its performance. If the system misses a page that actually underperforms (a false negative), the cost is the loss of click and traffic opportunities that could have been recaptured—that page continues to lose potential clicks because it never made it onto the review radar. Since reviewer capacity is limited, errors in both directions come with real costs: wasted time on one hand, and lost opportunities on the other. The cost of an incorrect recommendation can go two ways. If the system recommends a page that is actually fine (a false positive), the cost is the reviewer’s time wasted checking a page that doesn’t need to be changed—and in the worst-case scenario, a page that’s already good gets re-edited without a valid reason, risking a decline in its performance. If the system misses a page that actually underperforms (a false negative), the cost is the loss of click and traffic opportunities that could have been recaptured—that page continues to lose potential clicks because it never made it onto the review radar. Since reviewer capacity is limited, errors in both directions come with real costs: wasted time on one hand, and lost opportunities on the other.

The unit of analysis in this project is a single content page (content_id), based on its 90-day performance metrics (impressions, clicks, CTR, avg_position). The decision we aim to support: the content/SEO team has limited capacity to review pages one by one, so they need a priority list—which pages are most worth checking first because their CTR is significantly below expectations for their position. Actions that can be taken based on this output: reviewers open the top pages on the list and decide whether to rewrite the title or meta description, fix the snippet structure, or adjust the content to better match search intent—these are not automated actions, but recommendations that still require manual verification before being implemented. The cost of incorrect recommendations works both ways. If the system recommends a page that actually has no issues (false positive), the cost is the reviewer’s wasted time checking a page that doesn’t need to be changed—and in the worst-case scenar

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [17]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

import pandas as pd, numpy as np
df = pd.read_csv('data/raw/content_refresh_anonymized.csv')
print("Total rows:", df.shape[0])
visible = df[df['impressions_90d'] >= 150]
print("Rows after visibility filter >= 150", visible.shape[0])
ctr_by_pos = visible.groupby('position_tier')['ctr'].mean().sort_values(ascending=False)
print("CTR per position tier:")
print(ctr_by_pos)
tier_mean = visible.groupby('position_tier')['ctr'].transform("mean")
below_avg = visible['ctr'] < tier_mean
pct_below = below_avg.mean()
print(f"Pct pages below their tier average: {pct_below*100:.1f}%")
for factor in [0.3, 0.4, 0.5]:
  strict_gap = visible['ctr'] < (tier_mean * factor)
  print(f"Threshold {factor}: {strict_gap.mean()*100:.1f}% pages flagged")


Total rows: 30000
Rows after visibility filter >= 150 20993
CTR per position tier:
position_tier
page_1      0.343081
top_3       0.324360
striking    0.256924
page_3_5    0.140640
deep        0.054217
Name: ctr, dtype: float64
Pct pages below their tier average: 66.3%
Threshold 0.3: 34.8% pages flagged
Threshold 0.4: 39.5% pages flagged
Threshold 0.5: 45.4% pages flagged


From the results above, it is clear that expected CTR is highly dependent on search result position—the better the position, the higher the average CTR (ranging from 0.054 in “deep” to 0.343 in “page_1”). However, within each position tier, there are still pages whose CTR is well below the tier’s average—about 45.4% of visible pages fall below 50% of their tier’s average CTR. This indicates that the gap isn’t solely a matter of position, but that there are specific pages that underperform relative to their respective “classes.”

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

### Based on this analysis, I can claim: there is an observed gap between a page’s actual CTR and the expected CTR for its respective position tier—approximately 45.4% of visible pages show a CTR less than half of their tier’s average. This is a directional pattern that serves as a starting point for prioritizing reviews, not a final conclusion. The output of this work is decision-support in nature: it provides reviewers with a more targeted list of candidates rather than having to check every page one by one, but the final decision (whether a page truly needs to be changed, and how) remains in human hands. I cannot claim that a low CTR is definitely caused by poor titles or meta tags—it could be due to mismatched search intent, irrelevant content, or even just noise resulting from low volume despite filtering. I also can’t claim that fixing these pages will automatically increase CTR or traffic—that requires experimentation (such as A/B testing or controlled before-and-after studies), not just observational analysis like this. Finally, I’m not claiming to have proven anything about Google’s algorithm—everything observed here is purely patterns derived from the available data (impressions, clicks, positions), not how the search ranking system itself works.

Based on this analysis, I can claim: there is an observed gap between a page’s actual CTR and the expected CTR for its respective position tier—approximately 45.4% of visible pages show a CTR less than half of their tier’s average. This is a directional pattern that serves as a starting point for prioritizing reviews, not a final conclusion. The output of this work is decision-support in nature: it provides reviewers with a more targeted list of candidates rather than having to check every page one by one, but the final decision (whether a page truly needs to be changed, and how) remains in human hands. I cannot claim that a low CTR is definitely caused by poor titles or meta tags—it could be due to mismatched search intent, irrelevant content, or even just noise resulting from low volume despite filtering. I also can’t claim that fixing these pages will automatically increase CTR or traffic—that requires experimentation (such as A/B testing or controlled before-and-after studies), not

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.